src/dpdkstat.c

```diff
@@ -0,0 +1,775 @@
+  DEBUG("dpdkstat: %s: ports %d, xstats %d\n", __func__, temp_config.num_ports,
+        temp_config.num_xstats);
+
+  int shm_xstats_size = sizeof(dpdk_config_t) + (sizeof(struct rte_eth_xstats) *
+                        g_configuration->num_xstats);
+  DEBUG("=== SHM new size for %d xstats\n", g_configuration->num_xstats);
+
+  int err = dpdk_shm_cleanup();
+  if (err)
+    ERROR("dpdkstat: Error in shm_cleanup in %s\n", __func__);
+
+  err = dpdk_shm_init(shm_xstats_size);
+  if (err)
+    ERROR("dpdkstat: Error in shm_init in %s\n", __func__);
+
+  /* If the XML config() function has been run, dont re-initialize defaults */
+  if(!g_configured)
+    dpdk_config_init_default();
+
+  memcpy(g_configuration, &temp_config, sizeof(dpdk_config_t));
+  g_configuration->collectd_reinit_shm = 0;
+
+  return 0;
+}
+
+static int dpdk_init (void)
+{
+  int err = dpdk_shm_init(sizeof(dpdk_config_t));
+  if (err)
+    ERROR("dpdkstat: %s : error %d in shm_init()", __func__, err);
+
+  /* If the XML config() function has been run, dont re-initialize defaults */
+  if(!g_configured) {
+    dpdk_config_init_default();
+  }
+
+  plugin_register_complex_read (NULL, "dpdkstat", dpdk_read,
+                                g_configuration->interval, NULL);
+  return 0;
+}
+
+static int dpdk_helper_exit(int reset)
+{
+  g_configuration->helper_status = DPDK_HELPER_GRACEFUL_QUIT;
+  if (reset) {
+    g_configuration->eal_initialized = 0;
+    g_configuration->num_ports = 0;
+    memset(&g_configuration->xstats, 0, g_configuration->num_xstats* sizeof(struct rte_eth_xstats));
+    g_configuration->num_xstats = 0;
+    for (int i = 0; i < RTE_MAX_ETHPORTS; i++)
+      g_configuration->num_stats_in_port[i] = 0;
+  }
+  close(g_configuration->helper_pipes[1]);
+  int err = kill(g_configuration->helper_pid, SIGKILL);
+  if (err) {
+    ERROR("dpdkstat: error sending kill to helper: %s\n", strerror(errno));
+  }
+
+  return 0;
+}
+
+static int dpdk_helper_spawn(enum DPDK_HELPER_ACTION action)
+{
+  g_configuration->eal_initialized = 0;
+  g_configuration->helper_action = action;
+  /*
+   * Create a pipe for helper stdout back to collectd. This is necessary for
+   * logging EAL failures, as rte_eal_init() calls rte_panic().
+   */
+  if (g_configuration->helper_pipes[1]) {
+    DEBUG("dpdkstat: collectd closing helper pipe %d\n",
+          g_configuration->helper_pipes[1]);
+  } else {
+    DEBUG("dpdkstat: collectd helper pipe %d, not closing\n",
+          g_configuration->helper_pipes[1]);
+  }
+  if(pipe(g_configuration->helper_pipes) != 0) {
+    DEBUG("dpdkstat: Could not create helper pipe: %s\n", strerror(errno));
+    return -1;
+  }
+
+  int pipe0_flags = fcntl(g_configuration->helper_pipes[0], F_GETFL, 0);
+  int pipe1_flags = fcntl(g_configuration->helper_pipes[1], F_GETFL, 0);
+  if (pipe0_flags == -1 || pipe1_flags == -1) {
+    ERROR("dpdkstat: error setting up pipe flags: %s\n", strerror(errno));
```

Comment:
Continuing here at the very least required to initialize `pipe0_flags` and/or `pipe1_flags` to zero: -1 will mean "all flags" on most platforms, which results in an error from `fcntl()` at best and weird and hard to debug behavior at worst. If you chose to continue, please use a `WARNING()` instead of an error. `ERROR()` means "something went bad and I *cannot* go on" while `WARNING()` means "something went bad but I can carry on".

Please use `sstrerror()` (note the double "s") which is thread-safe. Please don't include a newline in the format string.